### PR TITLE
Added documentation for @KeepGeneratedSerializer feature

### LINKED ISSUE #2660: Add documentation for @KeepSerializer feature
Its KDoc should be updated, and a section in the guide should be added.

### PR DESCRIPTION
**Pull request by shanshin**
Resolves #2660

### CODE REVIEW

**Review by sandwwraith — CHANGES_REQUESTED** [review]
I think it is worth adding an example with `JsonTransformingSerializer(MyObj.generatedSerializer())` as well (see https://github.com/Kotlin/kotlinx.serialization/issues/1310) — likely to `json.md#json-transformations` section that will explain how to make a JsonTransformingSerializer a default one while delegating to generated serializer at the same time.

Also, `knitCheck` task fails.

**Comment by shanshin** [comment]
Waiting for `2.0.20` release

File: docs/json.md [hunk 1]
```diff
@@ -1260,6 +1261,51 @@ No class discriminator is added in the JSON output:
 
 <!--- TEST -->
 
+### Extending the behavior of the plugin generated serializer
+In some cases, it may be necessary to add additional serialization logic on top of the plugin generated logic.
+For example, it is necessary to add a preliminary modification of JSON elements or to add processing of unknown values of enums.
```
Comment (sandwwraith) [inline docs/json.md:1266]:
```suggestion
For example, to add a preliminary modification of JSON elements or to add processing of unknown values of enums.
```

File: docs/json.md [hunk 2]
```diff
@@ -1260,6 +1261,51 @@ No class discriminator is added in the JSON output:
...
+
+In this case, you should mark the serializable class with the [`@KeepGeneratedSerializer`][KeepGeneratedSerializer] annotation and get the generated serializer using the `generatedSerializer()` function.
```
Comment (sandwwraith) [inline docs/json.md:1268]:
```suggestion
In this case, you can mark the serializable class with the [`@KeepGeneratedSerializer`][KeepGeneratedSerializer] annotation and get the generated serializer using the `generatedSerializer()` function.
```

File: docs/json.md [hunk 3]
```diff
@@ -1260,6 +1261,51 @@ No class discriminator is added in the JSON output:
...
+
+An example of the simultaneous use of [JsonTransformingSerializer] and polymorphism:
```
Comment (sandwwraith) [inline docs/json.md:1270]:
```suggestion
Here is an example of the simultaneous use of [JsonTransformingSerializer] and polymorphism.
In this example, we use `transformDeserialize` function to rename `basic-name` key into `name` so it matches the `abstract val name` property from the `Project` supertype.
```

**Review by sandwwraith — APPROVED** [review]

**Comment by shanshin** [comment]
I think it would be safer to add https://github.com/Kotlin/kotlinx.serialization/pull/2758

File: core/commonMain/src/kotlinx/serialization/Annotations.kt [hunk 4]
```diff
@@ -330,13 +334,15 @@ public annotation class Polymorphic
...
  * Automatically generated serializer is available via `generatedSerializer()` function in companion object of serializable class.
  *
- * Generated serializers allow to use custom serializers on classes from which other serializable classes are inherited.
+ * Keeping generated serializers allow to use plugin generated serializer in inheritors even if custom serializer is specified.
+ *
+ * Used only with annotation [Serializable] with the specified argument [Serializable.with], e.g. `@Serializable(with=SomeSerializer::class)`.
  *
- * Used only with the [Serializable] annotation.
+ * Annotation is not allowed on classes involved in polymorphic serialization:
+ * interfaces, sealed classes, abstract classes, classes marked by [Polymorphic].
```
Comment (hfhbd) [inline core/commonMain/src/kotlinx/serialization/Annotations.kt:342]: I don't understand this sentence: `@KeepGeneratedSerializer` is not allowed with any inheritance, but your sample code uses a class with inherits from an interface. (see other comment)

File: docs/json.md [hunk 5]
```diff
@@ -1260,6 +1261,53 @@ No class discriminator is added in the JSON output:
...
+```kotlin
+@Serializable
+sealed class Project {
+    abstract val name: String
+}
+
+@KeepGeneratedSerializer
+@Serializable(with = BasicProjectSerializer::class)
+@SerialName("basic")
+data class BasicProject(override val name: String): Project()
```
Comment (hfhbd) [inline docs/json.md:1281]: Also, do you test your sample automatically? If not, can you add this code to your tests in #2758?
Background: I use Kotlin 2.0.20-RC with kx-serial 1.7.1, and I get this error:
```
java.lang.NullPointerException
	at kotlinx.serialization.SealedClassSerializer$special$$inlined$groupingBy$1.keyOf(_Collections.kt:1547)
```
I will also create an issue with a small reproducer.

File: core/commonMain/src/kotlinx/serialization/Annotations.kt (hunk 4 above)
Comment (shanshin) [reply core/commonMain/src/kotlinx/serialization/Annotations.kt:342]: It means that you cannot specify the `@KeepGeneratedSerializer ` interface on the interfaces, sealed/abstract classes, but it's ok to add this annotation add to the heirs

File: docs/json.md (hunk 5 above)
Comment (shanshin) [reply docs/json.md:1281]: > Also, do you test your sample automatically?

yes

> I will also create an issue with a small reproducer.

Thanks!

**Pull request merged**